pandas/core/arrays/arrow/array.py

```diff
@@ -1292,15 +1291,21 @@ def _mode(self: ArrowExtensionArrayT, dropna: bool = True) -> ArrowExtensionArra
             data = self._data
 
         modes = pc.mode(data, pc.count_distinct(data).as_py())
-        values = modes.field(0)
         counts = modes.field(1)
         # counts sorted descending i.e counts[0] = max
+        if not dropna and self._data.null_count > counts[0].as_py():
```

Comment:
i know you've wanted to keep as much of this "in pyarrow" as possible, but i find this hard to follow.  would it be that bad to just implement mode in terms of value_counts?

```
if not len(self):[...]
vcs = self.value_counts(dropna=dropna)
res_ser = vcs[vcs == vcs.max()].sort_index()
return res_ser.index._values
```

Reply:
makes sense, thanks for taking a look

Reply:
Seeing this comment here, I find it surprising that there is such a big difference between both. Especially because you are doing a count_distinct to include all unique values in the result of `mode`, so essentially making it equivalent to a value_counts. So if there is a big difference, that seems to indicate some performance issue in the implementation in arrow. 

Now, trying with the following, I see different results:

```
In [8]: arr = pa.array(np.random.randint(0, 1000, 1_000_000))

In [12]: %timeit pc.mode(arr, pc.count_distinct(arr).as_py())
12.7 ms ± 496 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [13]: %timeit pc.value_counts(arr)
8.96 ms ± 170 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
(it might depend a lot on the exact characteristics of the data, though, number of uniques vs total number of values, etc)

Reply:
This was a quick benchmark replacing the implementation to use `ArrowExtensionArray.value_counts`, not just comparing `pc.mode` vs `pc.value_counts` per se

```
% ipython
Python 3.8.15 | packaged by conda-forge | (default, Nov 22 2022, 08:53:40)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]: arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
15.9 ms ± 57.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

% git diff
+        vcs = self.value_counts(dropna=dropna)
+        res_ser = vcs[vcs == vcs.max()].sort_index()
+        return res_ser.index._values
+        # pa_type = self._data.type
+        # if pa.types.is_temporal(pa_type):
+        #     nbits = pa_type.bit_width
+        #     if nbits == 32:
+        #         data = self._data.cast(pa.int32())
+        #     elif nbits == 64:
+        #         data = self._data.cast(pa.int64())
+        #     else:
+        #         raise NotImplementedError(pa_type)
+        # else:
+        #     data = self._data
+        #
+        # modes = pc.mode(data, pc.count_distinct(data).as_py())
+        # counts = modes.field(1)
+        # # counts sorted descending i.e counts[0] = max
+        # if not dropna and self._data.null_count > counts[0].as_py():
+        #     return type(self)(pa.array([None], type=pa_type))
+        # mask = pc.equal(counts, counts[0])
+        # most_common = modes.field(0).filter(mask)
+        #
+        # if pa.types.is_temporal(pa_type):
+        #     most_common = most_common.cast(pa_type)
+        #
+        # if not dropna and self._data.null_count == counts[0].as_py():
+        #     most_common = pa.concat_arrays(
+        #         [most_common, pa.array([None], type=pa_type)]
+        #     )
+        #
+        # return type(self)(most_common)

% ipython
Python 3.8.15 | packaged by conda-forge | (default, Nov 22 2022, 08:53:40)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]: arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
54.7 ms ± 672 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
Then I suppose the overhead is coming from the extra work the `ArrowExtensionArray.value_counts` is doing? 

Doing it with pyarrow compute directly here in `_mode` might be faster and simpler (compared to current `_mode`)? Something like: 

```
res = pc.value_counts(self.data)
most_common = res.field("values").filter(pc.equal(res.field("counts"), pc.max(res.field("counts"))))
```

(this is still faster than calling `pc.mode` on your example data)

Reply:
Nice! Looks like `pc.value_counts` also has some benefits

1. Works for string and binary types
2. The result maintains the order of the original values.

So it's good to switch to using value_counts here